In [1]:
pip install google-cloud-bigquery

Note: you may need to restart the kernel to use updated packages.


### Loading both tables (kelmar and sok)

In [58]:
from google.cloud import bigquery

client = bigquery.Client(project="omes-solacc-citizenmatch-01-d")

kelmar_query = """
SELECT *
FROM `omes-solacc-citizenmatch-01-d.citizen_match.kelmar_staging_dataset`
"""

kelmar_df = client.query(kelmar_query).to_dataframe()

print("Kelmar shape:", kelmar_df.shape)

Kelmar shape: (5000, 14)


### Inspect SSN format in kelmar

In [59]:
import pandas as pd
import re

def inspect_ssn(df, ssn_col="SSN"):
    sample = df[df[ssn_col].notna()][ssn_col].head(50)

    results = pd.DataFrame({
        "SSN": sample,
        "length": sample.astype(str).apply(len),
        "has_dash": sample.astype(str).str.contains("-"),
        "has_space": sample.astype(str).str.contains(" "),
        "all_digits": sample.astype(str).str.match(r"^\d+$"),
        "masked": sample.astype(str).str.contains(r"X", case=False)
    })

    return results

inspect_ssn(kelmar_df)

,SSN,length,has_dash,has_space,all_digits,masked
0,467362640,9,False,False,True,False
1,443082421,9,False,False,True,False
2,446428255,9,False,False,True,False
3,538399548,9,False,False,True,False
4,444600938,9,False,False,True,False
5,448400682,9,False,False,True,False
6,448400682,9,False,False,True,False
7,448400682,9,False,False,True,False
8,448113304,9,False,False,True,False
9,241674629,9,False,False,True,False


### Checking ssn format in sok without loading full table

In [60]:
sok_query = """
SELECT SSN
FROM `omes-solacc-citizenmatch-01-d.citizen_match.sok_staging_dataset`
WHERE SSN IS NOT NULL
LIMIT 5
"""

sok_sample_df = client.query(sok_query).to_dataframe()

inspect_ssn(sok_sample_df)

,SSN,length,has_dash,has_space,all_digits,masked
0,025604287,9,False,False,True,False
1,387150031,9,False,False,True,False
2,484048388,9,False,False,True,False
3,509028032,9,False,False,True,False
4,558610846,9,False,False,True,False


### Encrypt ALL SOK

In [61]:
import time
import pandas as pd
from google.cloud import bigquery, dlp_v2

# -------------------
# Config
# -------------------
PROJECT_ID = "omes-solacc-citizenmatch-01-d"
LOCATION = "us-central1"
DATASET = "citizen_match"

SOK_SOURCE = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset"

# ✅ Deterministic template (v2)
DLP_TEMPLATE_NAME = "projects/omes-solacc-citizenmatch-01-d/locations/us-central1/deidentifyTemplates/3184577974128662341"

# ✅ NEW destination table (do not overwrite v1)
SOK_TOKEN_TABLE = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset_tokenized_v2"

# Chunk sizes
BQ_CHUNK_ROWS = 100_000
DLP_BATCH_SIZE = 200
MAX_RETRIES = 6

# -------------------
# Clients
# -------------------
bq = bigquery.Client(project=PROJECT_ID)
dlp = dlp_v2.DlpServiceClient()
parent = f"projects/{PROJECT_ID}/locations/{LOCATION}"

# -------------------
# DLP batch tokenization (TABLE item)
# -------------------
def tokenize_ssns_with_dlp_table(ssns, batch_size=DLP_BATCH_SIZE, max_retries=MAX_RETRIES):
    out_tokens = [None] * len(ssns)

    for start in range(0, len(ssns), batch_size):
        end = min(start + batch_size, len(ssns))
        batch = ssns[start:end]

        table_item = {
            "table": {
                "headers": [{"name": "SSN"}],
                "rows": [
                    {"values": [{"string_value": ("" if pd.isna(x) else str(x).strip())}]}
                    for x in batch
                ],
            }
        }

        delay = 1.0
        for _ in range(max_retries):
            try:
                resp = dlp.deidentify_content(
                    request={
                        "parent": parent,
                        "deidentify_template_name": DLP_TEMPLATE_NAME,
                        "item": table_item,
                    }
                )
                tokens = [r.values[0].string_value for r in resp.item.table.rows]
                out_tokens[start:end] = tokens
                break
            except Exception as e:
                msg = str(e)
                if "ResourceExhausted" in msg or "429" in msg:
                    time.sleep(delay)
                    delay *= 2
                    continue
                raise
        else:
            raise RuntimeError(f"Failed after retries for batch {start}:{end}")

    return out_tokens

# -------------------
# Keyset pagination on stable composite key:
# Primary: (Transaction_ID, DLN)
# Tie-breakers to never skip duplicates: (_data_file_date_, SSN)
# NOTE: Transaction_ID can be NULL -> COALESCE to '' so rows are still included.
# -------------------
def fetch_sok_chunk(last_txn=None, last_dln=None, last_file=None, last_ssn=None, limit=BQ_CHUNK_ROWS):
    where = "WHERE SSN IS NOT NULL AND DLN IS NOT NULL"

    if last_txn is not None:
        where += f"""
        AND (
          COALESCE(CAST(Transaction_ID AS STRING), '') > '{last_txn}'
          OR (
            COALESCE(CAST(Transaction_ID AS STRING), '') = '{last_txn}'
            AND (
              CAST(DLN AS STRING) > '{last_dln}'
              OR (
                CAST(DLN AS STRING) = '{last_dln}'
                AND (
                  COALESCE(CAST(_data_file_date_ AS STRING), '') > '{last_file}'
                  OR (
                    COALESCE(CAST(_data_file_date_ AS STRING), '') = '{last_file}'
                    AND CAST(SSN AS STRING) > '{last_ssn}'
                  )
                )
              )
            )
          )
        )
        """

    query = f"""
    SELECT
      COALESCE(CAST(Transaction_ID AS STRING), '') AS Transaction_ID,
      COALESCE(CAST(_data_file_date_ AS STRING), '') AS _data_file_date_,
      Transaction_Date,
      Transaction_Type,

      CAST(DLN AS STRING) AS DLN,
      CAST(SSN AS STRING) AS SSN,

      First_Name, Middle_Name, Last_Name, Suffix,
      Date_of_Birth,

      Residential_Address_Street, Residential_Address_Street_2,
      Residential_Address_Unit_Type, Residential_Address_Unit,
      Residential_Address_City, Residential_Address_State, Residential_Address_Zip,

      Mailing_Address_Street, Mailing_Address_Street_2,
      Mailing_Address_Unit_Type, Mailing_Address_Unit,
      Mailing_Address_City, Mailing_Address_State, Mailing_Address_Zip
    FROM `{SOK_SOURCE}`
    {where}
    ORDER BY
      COALESCE(CAST(Transaction_ID AS STRING), ''),
      CAST(DLN AS STRING),
      COALESCE(CAST(_data_file_date_ AS STRING), ''),
      CAST(SSN AS STRING)
    LIMIT {limit}
    """
    return bq.query(query).to_dataframe(create_bqstorage_client=True)

# -------------------
# Write helper
# -------------------
def write_chunk_to_bq(df, table_id, first_write=False):
    job_config = bigquery.LoadJobConfig(
        write_disposition="WRITE_TRUNCATE" if first_write else "WRITE_APPEND"
    )
    job = bq.load_table_from_dataframe(df, table_id, job_config=job_config)
    job.result()

# -------------------
# Main loop
# -------------------
last_txn = None
last_dln = ""
last_file = ""
last_ssn = ""
first_write = True
total_written = 0

while True:
    chunk = fetch_sok_chunk(
        last_txn=last_txn,
        last_dln=last_dln,
        last_file=last_file,
        last_ssn=last_ssn,
        limit=BQ_CHUNK_ROWS
    )

    if chunk.empty:
        break

    # Tokenize SSN for this chunk
    chunk["SSN"] = chunk["SSN"].astype("string").str.strip()
    chunk["ssn_token"] = tokenize_ssns_with_dlp_table(chunk["SSN"].tolist(), batch_size=DLP_BATCH_SIZE)

    # Compute last4, BUT keep SSN until after cursor update
    chunk["ssn_last4"] = chunk["SSN"].str.replace(r"\D", "", regex=True).str[-4:]

    # Advance cursor using the last row of this chunk (uses the ORDER BY fields)
    last_txn = chunk["Transaction_ID"].iloc[-1]
    last_dln = chunk["DLN"].iloc[-1]
    last_file = chunk["_data_file_date_"].iloc[-1]
    last_ssn = chunk["SSN"].iloc[-1]

    # Drop raw SSN for safety before writing
    chunk = chunk.drop(columns=["SSN"])

    # Write to BigQuery
    write_chunk_to_bq(chunk, SOK_TOKEN_TABLE, first_write=first_write)
    first_write = False

    total_written += len(chunk)

    print(
        f"Wrote {len(chunk):,} rows | Total written: {total_written:,} | "
        f"last_txn={last_txn} last_dln={last_dln} last_file={last_file} last_ssn={last_ssn}"
    )

print(f"✅ Done. Total rows written to {SOK_TOKEN_TABLE}: {total_written:,}")

Wrote 100,000 rows | Total written: 100,000 | last_txn= last_dln=990309346 last_file=2026-02-13 last_ssn=509067702
Wrote 100,000 rows | Total written: 200,000 | last_txn= last_dln=A088047871 last_file=2026-02-13 last_ssn=103861312
Wrote 100,000 rows | Total written: 300,000 | last_txn= last_dln=A088166410 last_file=2026-02-13 last_ssn=764561569
Wrote 100,000 rows | Total written: 400,000 | last_txn= last_dln=B082362653 last_file=2026-02-13 last_ssn=445225914
Wrote 100,000 rows | Total written: 500,000 | last_txn= last_dln=C082901647 last_file=2026-02-13 last_ssn=447968892
Wrote 100,000 rows | Total written: 600,000 | last_txn= last_dln=D081869008 last_file=2026-02-13 last_ssn=439608171
Wrote 100,000 rows | Total written: 700,000 | last_txn= last_dln=E021453435 last_file=2026-02-13 last_ssn=645170946
Wrote 100,000 rows | Total written: 800,000 | last_txn= last_dln=E084091672 last_file=2026-02-13 last_ssn=447130158
Wrote 100,000 rows | Total written: 900,000 | last_txn= last_dln=F0815869

### Encrypt ALL Kelmar

In [62]:
import time
import pandas as pd
from google.cloud import bigquery, dlp_v2

# -------------------
# Config
# -------------------
PROJECT_ID = "omes-solacc-citizenmatch-01-d"
LOCATION = "us-central1"
DATASET = "citizen_match"

# ✅ UPDATE THIS to your Kelmar staging table name in BigQuery
KELMAR_SOURCE_TABLE = f"{PROJECT_ID}.{DATASET}.kelmar_staging_dataset"

# Destination table for tokenized Kelmar
KELMAR_TOKEN_TABLE = f"{PROJECT_ID}.{DATASET}.kelmar_staging_dataset_tokenized_v1"

# ✅ Deterministic template (same one used for SOK)
DLP_TEMPLATE_NAME = "projects/omes-solacc-citizenmatch-01-d/locations/us-central1/deidentifyTemplates/3184577974128662341"

DLP_BATCH_SIZE = 200
MAX_RETRIES = 6

# -------------------
# Clients
# -------------------
bq = bigquery.Client(project=PROJECT_ID)
dlp = dlp_v2.DlpServiceClient()
parent = f"projects/{PROJECT_ID}/locations/{LOCATION}"

# -------------------
# DLP batch tokenization (TABLE item)
# -------------------
def tokenize_ssns_with_dlp_table(ssns, batch_size=DLP_BATCH_SIZE, max_retries=MAX_RETRIES):
    out_tokens = [None] * len(ssns)

    for start in range(0, len(ssns), batch_size):
        end = min(start + batch_size, len(ssns))
        batch = ssns[start:end]

        table_item = {
            "table": {
                "headers": [{"name": "SSN"}],
                "rows": [
                    {"values": [{"string_value": ("" if pd.isna(x) else str(x).strip())}]}
                    for x in batch
                ],
            }
        }

        delay = 1.0
        for _ in range(max_retries):
            try:
                resp = dlp.deidentify_content(
                    request={
                        "parent": parent,
                        "deidentify_template_name": DLP_TEMPLATE_NAME,
                        "item": table_item,
                    }
                )
                tokens = [r.values[0].string_value for r in resp.item.table.rows]
                out_tokens[start:end] = tokens
                break
            except Exception as e:
                msg = str(e)
                if "ResourceExhausted" in msg or "429" in msg:
                    time.sleep(delay)
                    delay *= 2
                    continue
                raise
        else:
            raise RuntimeError(f"Failed after retries for batch {start}:{end}")

    return out_tokens

# -------------------
# Read Kelmar (all rows; ~5k)
# -------------------
kelmar_query = f"""
SELECT
  OwnerID,
  PropertyID,
  NameLast,
  NameFirst,
  NameMiddle,
  Address1,
  Address2,
  Address3,
  City,
  State,
  Zip,
  CAST(SSN AS STRING) AS SSN,
  BirthDT,
  CashValue
FROM `{KELMAR_SOURCE_TABLE}`
"""

kelmar_df = bq.query(kelmar_query).to_dataframe(create_bqstorage_client=True)
print("Loaded Kelmar:", kelmar_df.shape)

# -------------------
# Tokenize SSN (deterministic)
# -------------------
kelmar_df["SSN"] = kelmar_df["SSN"].astype("string").str.strip()
kelmar_df["ssn_token"] = tokenize_ssns_with_dlp_table(kelmar_df["SSN"].tolist(), batch_size=DLP_BATCH_SIZE)
kelmar_df["ssn_last4"] = kelmar_df["SSN"].str.replace(r"\D", "", regex=True).str[-4:]

# Drop raw SSN for safety
kelmar_df = kelmar_df.drop(columns=["SSN"])

# -------------------
# Write tokenized Kelmar table (replace)
# -------------------
job = bq.load_table_from_dataframe(
    kelmar_df,
    KELMAR_TOKEN_TABLE,
    job_config=bigquery.LoadJobConfig(write_disposition="WRITE_TRUNCATE"),
)
job.result()

print(f"✅ Done. Wrote {len(kelmar_df):,} rows to {KELMAR_TOKEN_TABLE}")
kelmar_df.head()

Loaded Kelmar: (5000, 14)
✅ Done. Wrote 5,000 rows to omes-solacc-citizenmatch-01-d.citizen_match.kelmar_staging_dataset_tokenized_v1


,OwnerID,PropertyID,NameLast,NameFirst,NameMiddle,Address1,Address2,Address3,City,State,Zip,BirthDT,CashValue,ssn_token,ssn_last4
0,13561708,13213300,WHITTEN,ROBERT D,GST,,,,Oklahoma City,OK,<NA>,NaT,41.34,SSN_TOKEN(36):AYBeT6QPOxXWndWNvdzeEV8b3CMMeCyg...,2640
1,13570556,13222067,HORN,BRADEN,JOSEPH,414 LITCHFIELD LN,,,NORMAN,OK,<NA>,NaT,61.02,SSN_TOKEN(36):Ady2DEfslzSYMaNqAB+ZTzg/kFNlapQY...,2421
2,12781948,12475205,CANTRELL,OGDON,KAYE,19525 SPORTSMANS RD,,,EDMOND,OK,73003,NaT,40.00,SSN_TOKEN(36):AWOnPyoge97Zf/uDzKqJIXzKZ0xl/eSi...,8255
3,13572605,13224114,WILLIAMS,GORDON,ERNEST,310 N 103 RD ST,,,WASHITA,OK,73005,1998-04-24,500.00,SSN_TOKEN(36):AbgXnsK68spzlI8MkA4xexNwo5HyLTwO...,9548
4,12761622,12458648,ODOR,CAROLYN,MONNET,15800 N WESTERN AVE,,,EDMOND,OK,73013,NaT,68.04,SSN_TOKEN(36):ARB8E5VXGE2r08kdj7hF9u26cPdtB29H...,0938


### Cleaning SOK table using our tokenized table

In [63]:
import re
import unicodedata
import pandas as pd
from google.cloud import bigquery

# -------------------
# CONFIG
# -------------------
PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

SOK_SOURCE = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset_tokenized_v2"
SOK_CLEAN  = f"{PROJECT_ID}.{DATASET}.sok_clean_v1"

bq = bigquery.Client(project=PROJECT_ID)

# -------------------
# CLEANERS (exactly yours)
# -------------------
def rm_accents(x: str) -> str:
    return unicodedata.normalize("NFKD", x).encode("ASCII", "ignore").decode("ASCII")

def collapse_ws(s: str) -> str:
    return re.sub(r"\s+", " ", s).strip()

PUNCT_NAME_DROP = r"[^A-Z'\-\s]"
PUNCT_ADDR_KEEP = r"[^A-Z0-9'\-\s]"
TITLE_RE = r"\b(MR|MRS|MS|MISS|DR|PROF|REV|SRGT|SGT|OFFICER|ATTY|ATTORNEY|JUDGE)\b"

DIR_MAP = {"NORTH":"N","SOUTH":"S","EAST":"E","WEST":"W","NORTHEAST":"NE","NORTHWEST":"NW","SOUTHEAST":"SE","SOUTHWEST":"SW"}
STREET_MAP = {"STREET":"ST","AVENUE":"AVE","ROAD":"RD","LANE":"LN","DRIVE":"DR","BOULEVARD":"BLVD","COURT":"CT","PLACE":"PL","PARKWAY":"PKWY","CIRCLE":"CIR","HIGHWAY":"HWY"}

def clean_name(val):
    if val is None: return None
    v = str(val).strip()
    if v == "": return None
    v = rm_accents(v).upper()
    v = re.sub(TITLE_RE, "", v)
    v = re.sub(PUNCT_NAME_DROP, " ", v)
    v = collapse_ws(v)
    return v or None

def clean_state(val):
    if val is None: return None
    v = rm_accents(str(val)).upper()
    v = re.sub(r"[^A-Z]", "", v)[:2]
    return v or None

def clean_zip(val):
    if val is None: return None
    digits = re.sub(r"\D", "", str(val))[:5]
    return digits if digits else None

def clean_street(val1, val2=None):
    parts = [val1, val2] if val2 is not None else [val1]
    v = " ".join([str(p).strip() for p in parts if p is not None and str(p).strip() != ""])
    if v.strip() == "": return None
    v = rm_accents(v).upper()
    v = re.sub(r"\bP\.?\s*O\.?\s*B(?:OX)?\b", "PO BOX", v)
    v = re.sub(r"\b(APARTMENT|APT)\b", "APT", v)
    v = re.sub(r"\b(SUITE|STE)\b", "STE", v)
    v = re.sub(r"#\s*([A-Z0-9\-]+)", r"APT \1", v)
    v = re.sub(r"[.,]", "", v)
    v = re.sub(
        r"\b(NORTHWEST|NORTHEAST|SOUTHWEST|SOUTHEAST|NORTH|SOUTH|EAST|WEST)\b",
        lambda m: DIR_MAP[m.group(0)],
        v
    )
    st_pat = r"\b(" + "|".join(map(re.escape, STREET_MAP.keys())) + r")\b"
    v = re.sub(st_pat, lambda m: STREET_MAP[m.group(0)], v)
    v = re.sub(PUNCT_ADDR_KEEP, " ", v)
    v = collapse_ws(v)
    return v or None

# -------------------
# LOAD TOKENIZED SOK
# -------------------
df = bq.query(f"SELECT * FROM `{SOK_SOURCE}`").to_dataframe(create_bqstorage_client=True)
print("Loaded SOK tokenized:", df.shape)

# -------------------
# CLEAN NAMES
# -------------------
df["first_name_clean"]  = df["First_Name"].map(clean_name)
df["middle_name_clean"] = df["Middle_Name"].map(clean_name)
df["last_name_clean"]   = df["Last_Name"].map(clean_name)

df["full_name_clean"] = (
    df["first_name_clean"].fillna("") + " " +
    df["middle_name_clean"].fillna("") + " " +
    df["last_name_clean"].fillna("")
).map(lambda x: collapse_ws(x) if x.strip() != "" else None)

# -------------------
# CLEAN ADDRESS
# -------------------
df["street_clean"] = df.apply(
    lambda r: clean_street(r["Residential_Address_Street"], r["Residential_Address_Street_2"]),
    axis=1
)

df["city_clean"]  = df["Residential_Address_City"].map(clean_name)
df["state_clean"] = df["Residential_Address_State"].map(clean_state)
df["zip_clean"]   = df["Residential_Address_Zip"].map(clean_zip)

# -------------------
# WRITE CLEAN TABLE
# -------------------
job = bq.load_table_from_dataframe(
    df,
    SOK_CLEAN,
    job_config=bigquery.LoadJobConfig(write_disposition="WRITE_TRUNCATE"),
)
job.result()

print("✅ SOK cleaned table created:", SOK_CLEAN)

Loaded SOK tokenized: (4092400, 26)
✅ SOK cleaned table created: omes-solacc-citizenmatch-01-d.citizen_match.sok_clean_v1


### Cleaning Kelmar table using our tokenized table

In [67]:
import re
import unicodedata
import pandas as pd
from google.cloud import bigquery

# -------------------
# CONFIG
# -------------------
PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

KELMAR_SOURCE = f"{PROJECT_ID}.{DATASET}.kelmar_staging_dataset_tokenized_v1"
KELMAR_CLEAN  = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"

bq = bigquery.Client(project=PROJECT_ID)

# -------------------
# CLEANERS (same as SOK)
# -------------------
def rm_accents(x: str) -> str:
    return unicodedata.normalize("NFKD", x).encode("ASCII", "ignore").decode("ASCII")

def collapse_ws(s: str) -> str:
    return re.sub(r"\s+", " ", s).strip()

PUNCT_NAME_DROP = r"[^A-Z'\-\s]"
PUNCT_ADDR_KEEP = r"[^A-Z0-9'\-\s]"
TITLE_RE = r"\b(MR|MRS|MS|MISS|DR|PROF|REV|SRGT|SGT|OFFICER|ATTY|ATTORNEY|JUDGE)\b"

DIR_MAP = {"NORTH":"N","SOUTH":"S","EAST":"E","WEST":"W","NORTHEAST":"NE","NORTHWEST":"NW","SOUTHEAST":"SE","SOUTHWEST":"SW"}
STREET_MAP = {"STREET":"ST","AVENUE":"AVE","ROAD":"RD","LANE":"LN","DRIVE":"DR","BOULEVARD":"BLVD","COURT":"CT","PLACE":"PL","PARKWAY":"PKWY","CIRCLE":"CIR","HIGHWAY":"HWY"}

def clean_name(val):
    if val is None: return None
    v = str(val).strip()
    if v == "": return None
    v = rm_accents(v).upper()
    v = re.sub(TITLE_RE, "", v)
    v = re.sub(PUNCT_NAME_DROP, " ", v)
    v = collapse_ws(v)
    return v or None

def clean_state(val):
    if val is None: return None
    v = rm_accents(str(val)).upper()
    v = re.sub(r"[^A-Z]", "", v)[:2]
    return v or None

def clean_zip(val):
    if val is None: return None
    digits = re.sub(r"\D", "", str(val))[:5]
    return digits if digits else None

def clean_street(val1, val2=None, val3=None):
    parts = [val1, val2, val3]
    v = " ".join([str(p).strip() for p in parts if p is not None and str(p).strip() != ""])
    if v.strip() == "": return None
    v = rm_accents(v).upper()
    v = re.sub(r"\bP\.?\s*O\.?\s*B(?:OX)?\b", "PO BOX", v)
    v = re.sub(r"\b(APARTMENT|APT)\b", "APT", v)
    v = re.sub(r"\b(SUITE|STE)\b", "STE", v)
    v = re.sub(r"#\s*([A-Z0-9\-]+)", r"APT \1", v)
    v = re.sub(r"[.,]", "", v)
    v = re.sub(
        r"\b(NORTHWEST|NORTHEAST|SOUTHWEST|SOUTHEAST|NORTH|SOUTH|EAST|WEST)\b",
        lambda m: DIR_MAP[m.group(0)],
        v
    )
    st_pat = r"\b(" + "|".join(map(re.escape, STREET_MAP.keys())) + r")\b"
    v = re.sub(st_pat, lambda m: STREET_MAP[m.group(0)], v)
    v = re.sub(PUNCT_ADDR_KEEP, " ", v)
    v = collapse_ws(v)
    return v or None

# -------------------
# LOAD TOKENIZED KELMAR
# -------------------
df = bq.query(f"SELECT * FROM `{KELMAR_SOURCE}`").to_dataframe()
print("Loaded Kelmar tokenized:", df.shape)

# -------------------
# CLEAN NAMES
# -------------------
df["first_name_clean"]  = df["NameFirst"].map(clean_name)
df["middle_name_clean"] = df["NameMiddle"].map(clean_name)
df["last_name_clean"]   = df["NameLast"].map(clean_name)

df["full_name_clean"] = (
    df["first_name_clean"].fillna("") + " " +
    df["middle_name_clean"].fillna("") + " " +
    df["last_name_clean"].fillna("")
).map(lambda x: collapse_ws(x) if x.strip() != "" else None)

# -------------------
# CLEAN ADDRESS
# -------------------
df["street_clean"] = df.apply(
    lambda r: clean_street(r["Address1"], r["Address2"], r["Address3"]),
    axis=1
)

df["city_clean"]  = df["City"].map(clean_name)
df["state_clean"] = df["State"].map(clean_state)
df["zip_clean"]   = df["Zip"].map(clean_zip)

# -------------------
# WRITE CLEAN TABLE
# -------------------
job = bq.load_table_from_dataframe(
    df,
    KELMAR_CLEAN,
    job_config=bigquery.LoadJobConfig(write_disposition="WRITE_TRUNCATE"),
)
job.result()

print("✅ Kelmar cleaned table created:", KELMAR_CLEAN)

Loaded Kelmar tokenized: (5000, 15)
✅ Kelmar cleaned table created: omes-solacc-citizenmatch-01-d.citizen_match.kelmar_clean_v1


### Deterministic SSN Match (Clean Tables)

In [68]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

KELMAR = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"
SOK = f"{PROJECT_ID}.{DATASET}.sok_clean_v1"

OUTPUT = f"{PROJECT_ID}.{DATASET}.ssn_deterministic_matches_v1"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
CREATE OR REPLACE TABLE `{OUTPUT}` AS
SELECT
    -- Kelmar identifiers
    k.OwnerID,
    k.PropertyID,
    k.full_name_clean AS kelmar_name,
    k.BirthDT,
    k.CashValue,

    -- SOK identifiers
    s.DLN,
    s.full_name_clean AS sok_name,
    s._data_file_date_,
    s.Transaction_ID,
    s.Transaction_Date,
    s.Transaction_Type,

    -- shared
    k.ssn_token

FROM `{KELMAR}` k
JOIN `{SOK}` s
ON k.ssn_token = s.ssn_token
"""

bq.query(query).result()

print("✅ Deterministic match table created:", OUTPUT)

✅ Deterministic match table created: omes-solacc-citizenmatch-01-d.citizen_match.ssn_deterministic_matches_v1


In [69]:
count_q = f"""
SELECT COUNT(*) AS match_count
FROM `{OUTPUT}`
"""

print(bq.query(count_q).to_dataframe())

   match_count
0         3510


### We now isolate:

✅ Matched Kelmar (done)

❌ Unmatched Kelmar → goes to fuzzy stage

❌ SOK SSN-null + unmatched SSN-present

### Identify Unmatched Kelmar

In [70]:
UNMATCHED = f"{PROJECT_ID}.{DATASET}.kelmar_unmatched_v1"

query = f"""
CREATE OR REPLACE TABLE `{UNMATCHED}` AS
SELECT *
FROM `{PROJECT_ID}.{DATASET}.kelmar_clean_v1`
WHERE ssn_token NOT IN (
    SELECT DISTINCT ssn_token
    FROM `{PROJECT_ID}.{DATASET}.ssn_deterministic_matches_v1`
)
"""

bq.query(query).result()

print("✅ Unmatched Kelmar table created:", UNMATCHED)

✅ Unmatched Kelmar table created: omes-solacc-citizenmatch-01-d.citizen_match.kelmar_unmatched_v1


In [71]:
print(
    bq.query(f"SELECT COUNT(*) AS unmatched_count FROM `{UNMATCHED}`")
    .to_dataframe()
)

   unmatched_count
0             1516


### Next Step: Build the Correct Fuzzy Candidate Pool

Before we match anything fuzzily, we must define:

What SOK rows are eligible for fuzzy matching?

There are two possible groups:

Group A — SSN NULL rows (376,344)

These never had SSN. Must go to fuzzy.

Group B — SSN-present rows that did NOT match deterministically

These may represent:

SSN typo in Kelmar

SSN mismatch

Token issue

Identity fraud edge cases

We must decide whether to include them in fuzzy stage.

In [72]:
matched_sok_query = f"""
SELECT COUNT(DISTINCT DLN) AS matched_sok_rows
FROM `{PROJECT_ID}.{DATASET}.ssn_deterministic_matches_v1`
"""
print(bq.query(matched_sok_query).to_dataframe())

   matched_sok_rows
0              2344


### Deterministic SSN Match Summary

Kelmar matched rows: 3,510

Distinct SOK DLNs matched: 2,344

That means:

Some SOK DLNs matched multiple Kelmar properties.

This is expected because:

One citizen can have multiple properties.

Kelmar is property-level.

SOK is person-level (DLN).

So this distribution makes sense.

### Clean SSN NULL SOK Rows

In [74]:
SOK_NULL_SOURCE = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset"
SOK_NULL_CLEAN  = f"{PROJECT_ID}.{DATASET}.sok_ssn_null_clean_v1"

query = f"""
CREATE OR REPLACE TABLE `{SOK_NULL_CLEAN}` AS
SELECT *
FROM `{SOK_NULL_SOURCE}`
WHERE SSN IS NULL
"""

bq.query(query).result()

print("✅ SSN-null raw subset created:", SOK_NULL_CLEAN)

✅ SSN-null raw subset created: omes-solacc-citizenmatch-01-d.citizen_match.sok_ssn_null_clean_v1


In [75]:
import re
import unicodedata
import pandas as pd
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

SOK_NULL_SOURCE = f"{PROJECT_ID}.{DATASET}.sok_ssn_null_clean_v1"
SOK_NULL_CLEAN  = f"{PROJECT_ID}.{DATASET}.sok_ssn_null_clean_v1"

bq = bigquery.Client(project=PROJECT_ID)

# ---------- cleaners (exact same as before) ----------
def rm_accents(x: str) -> str:
    return unicodedata.normalize("NFKD", x).encode("ASCII", "ignore").decode("ASCII")

def collapse_ws(s: str) -> str:
    return re.sub(r"\s+", " ", s).strip()

PUNCT_NAME_DROP = r"[^A-Z'\-\s]"
PUNCT_ADDR_KEEP = r"[^A-Z0-9'\-\s]"
TITLE_RE = r"\b(MR|MRS|MS|MISS|DR|PROF|REV|SRGT|SGT|OFFICER|ATTY|ATTORNEY|JUDGE)\b"

DIR_MAP = {"NORTH":"N","SOUTH":"S","EAST":"E","WEST":"W","NORTHEAST":"NE","NORTHWEST":"NW","SOUTHEAST":"SE","SOUTHWEST":"SW"}
STREET_MAP = {"STREET":"ST","AVENUE":"AVE","ROAD":"RD","LANE":"LN","DRIVE":"DR","BOULEVARD":"BLVD","COURT":"CT","PLACE":"PL","PARKWAY":"PKWY","CIRCLE":"CIR","HIGHWAY":"HWY"}

def clean_name(val):
    if val is None: return None
    v = str(val).strip()
    if v == "": return None
    v = rm_accents(v).upper()
    v = re.sub(TITLE_RE, "", v)
    v = re.sub(PUNCT_NAME_DROP, " ", v)
    v = collapse_ws(v)
    return v or None

def clean_state(val):
    if val is None: return None
    v = rm_accents(str(val)).upper()
    v = re.sub(r"[^A-Z]", "", v)[:2]
    return v or None

def clean_zip(val):
    if val is None: return None
    digits = re.sub(r"\D", "", str(val))[:5]
    return digits if digits else None

def clean_street(val1, val2=None):
    parts = [val1, val2]
    v = " ".join([str(p).strip() for p in parts if p is not None and str(p).strip() != ""])
    if v.strip() == "": return None
    v = rm_accents(v).upper()
    v = re.sub(r"\bP\.?\s*O\.?\s*B(?:OX)?\b", "PO BOX", v)
    v = re.sub(r"\b(APARTMENT|APT)\b", "APT", v)
    v = re.sub(r"\b(SUITE|STE)\b", "STE", v)
    v = re.sub(r"#\s*([A-Z0-9\-]+)", r"APT \1", v)
    v = re.sub(r"[.,]", "", v)
    v = re.sub(
        r"\b(NORTHWEST|NORTHEAST|SOUTHWEST|SOUTHEAST|NORTH|SOUTH|EAST|WEST)\b",
        lambda m: DIR_MAP[m.group(0)],
        v
    )
    st_pat = r"\b(" + "|".join(map(re.escape, STREET_MAP.keys())) + r")\b"
    v = re.sub(st_pat, lambda m: STREET_MAP[m.group(0)], v)
    v = re.sub(PUNCT_ADDR_KEEP, " ", v)
    v = collapse_ws(v)
    return v or None

# ---------- load ----------
df = bq.query(f"SELECT * FROM `{SOK_NULL_SOURCE}`").to_dataframe(create_bqstorage_client=True)
print("Loaded SSN-null rows:", df.shape)

# ---------- clean ----------
df["first_name_clean"]  = df["First_Name"].map(clean_name)
df["middle_name_clean"] = df["Middle_Name"].map(clean_name)
df["last_name_clean"]   = df["Last_Name"].map(clean_name)

df["full_name_clean"] = (
    df["first_name_clean"].fillna("") + " " +
    df["middle_name_clean"].fillna("") + " " +
    df["last_name_clean"].fillna("")
).map(lambda x: collapse_ws(x) if x.strip() != "" else None)

df["street_clean"] = df.apply(
    lambda r: clean_street(r["Residential_Address_Street"], r["Residential_Address_Street_2"]),
    axis=1
)

df["city_clean"]  = df["Residential_Address_City"].map(clean_name)
df["state_clean"] = df["Residential_Address_State"].map(clean_state)
df["zip_clean"]   = df["Residential_Address_Zip"].map(clean_zip)

# ---------- write back ----------
job = bq.load_table_from_dataframe(
    df,
    SOK_NULL_CLEAN,
    job_config=bigquery.LoadJobConfig(write_disposition="WRITE_TRUNCATE"),
)
job.result()

print("✅ SSN-null rows cleaned successfully")

Loaded SSN-null rows: (376344, 58)
✅ SSN-null rows cleaned successfully


### Final Fuzzy Candidate Pool

In [77]:
FUZZY_POOL = f"{PROJECT_ID}.{DATASET}.sok_fuzzy_pool_v1"

query = f"""
CREATE OR REPLACE TABLE `{FUZZY_POOL}` AS

-- 1️⃣ SSN NULL cleaned rows
SELECT
    DLN,
    first_name_clean,
    middle_name_clean,
    last_name_clean,
    full_name_clean,
    street_clean,
    city_clean,
    state_clean,
    zip_clean,
    Date_of_Birth,
    NULL AS ssn_token
FROM `{PROJECT_ID}.{DATASET}.sok_ssn_null_clean_v1`

UNION ALL

-- 2️⃣ SSN-present cleaned rows NOT deterministically matched
SELECT
    s.DLN,
    s.first_name_clean,
    s.middle_name_clean,
    s.last_name_clean,
    s.full_name_clean,
    s.street_clean,
    s.city_clean,
    s.state_clean,
    s.zip_clean,
    s.Date_of_Birth,
    s.ssn_token
FROM `{PROJECT_ID}.{DATASET}.sok_clean_v1` s
LEFT JOIN `{PROJECT_ID}.{DATASET}.ssn_deterministic_matches_v1` m
ON s.ssn_token = m.ssn_token
WHERE m.ssn_token IS NULL
"""

bq.query(query).result()

print("✅ Fuzzy candidate pool created:", FUZZY_POOL)

✅ Fuzzy candidate pool created: omes-solacc-citizenmatch-01-d.citizen_match.sok_fuzzy_pool_v1


In [78]:
print(
    bq.query(f"SELECT COUNT(*) AS fuzzy_pool_count FROM `{FUZZY_POOL}`")
    .to_dataframe()
)

   fuzzy_pool_count
0           4466374


### Prepare Kelmar Fuzzy Table (Minimal Schema)

In [79]:
KELMAR_FUZZY = f"{PROJECT_ID}.{DATASET}.kelmar_fuzzy_v1"

query = f"""
CREATE OR REPLACE TABLE `{KELMAR_FUZZY}` AS
SELECT
    OwnerID,
    PropertyID,
    first_name_clean,
    middle_name_clean,
    last_name_clean,
    full_name_clean,
    street_clean,
    city_clean,
    state_clean,
    zip_clean,
    BirthDT
FROM `{PROJECT_ID}.{DATASET}.kelmar_unmatched_v1`
"""

bq.query(query).result()

print("✅ Kelmar fuzzy table created:", KELMAR_FUZZY)

✅ Kelmar fuzzy table created: omes-solacc-citizenmatch-01-d.citizen_match.kelmar_fuzzy_v1


In [80]:
print(
    bq.query(f"SELECT COUNT(*) AS kelmar_fuzzy_count FROM `{KELMAR_FUZZY}`")
    .to_dataframe()
)

   kelmar_fuzzy_count
0                1516


### Block Strategy 1
Exact ZIP + Exact Last Name + Exact DOB

### Candidate Pair Volume

In [81]:
BLOCK1 = f"{PROJECT_ID}.{DATASET}.fuzzy_block1_candidates_v1"

query = f"""
CREATE OR REPLACE TABLE `{BLOCK1}` AS
SELECT
    k.OwnerID,
    k.PropertyID,
    k.full_name_clean AS kelmar_name,
    k.street_clean AS kelmar_street,
    k.city_clean AS kelmar_city,
    k.state_clean AS kelmar_state,
    k.zip_clean AS kelmar_zip,
    k.BirthDT,

    s.DLN,
    s.full_name_clean AS sok_name,
    s.street_clean AS sok_street,
    s.city_clean AS sok_city,
    s.state_clean AS sok_state,
    s.zip_clean AS sok_zip,
    s.Date_of_Birth

FROM `{PROJECT_ID}.{DATASET}.kelmar_fuzzy_v1` k
JOIN `{PROJECT_ID}.{DATASET}.sok_fuzzy_pool_v1` s
ON k.zip_clean = s.zip_clean
AND k.last_name_clean = s.last_name_clean
AND k.BirthDT = s.Date_of_Birth
"""

bq.query(query).result()

print("✅ Block 1 candidate table created:", BLOCK1)

✅ Block 1 candidate table created: omes-solacc-citizenmatch-01-d.citizen_match.fuzzy_block1_candidates_v1


In [82]:
print(
    bq.query(f"SELECT COUNT(*) AS candidate_count FROM `{BLOCK1}`")
    .to_dataframe()
)

   candidate_count
0               20


### Score Block 1 Candidates

We will compute:

First name similarity

Full name similarity

Street similarity

### Pull Block 1 Candidates into dataframe

In [83]:
import pandas as pd

BLOCK1 = f"{PROJECT_ID}.{DATASET}.fuzzy_block1_candidates_v1"

block1_df = bq.query(f"SELECT * FROM `{BLOCK1}`").to_dataframe()
print("Loaded Block 1 candidates:", block1_df.shape)
block1_df.head()

Loaded Block 1 candidates: (20, 15)


,OwnerID,PropertyID,kelmar_name,kelmar_street,kelmar_city,kelmar_state,kelmar_zip,BirthDT,DLN,sok_name,sok_street,sok_city,sok_state,sok_zip,Date_of_Birth
0,12771949,12466462,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,507 W MAIN ST,CALUMET,OK,73014,1960-05-07
1,12771950,12466480,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,507 W MAIN ST,CALUMET,OK,73014,1960-05-07
2,13439549,13099416,ALLEN H MOORE,2205 WINDSONG,EDMOND,OK,73034,1971-12-17,V081149038,ALLEN H MOORE,2205 WINDSONG,EDMOND,OK,73034,1971-12-17
3,13440139,13100006,ROBERT N ALLEN,2101 N FONSHILL AVE,OKLAHOMA CITY,OK,73111,1946-01-22,447449471,ROBERT NELSON ALLEN,2101 N FONSHILL AVE,OKLA CITY,OK,73111,1946-01-22
4,12897104,12577135,NANCY COUCH,6612 RANDI RD,OKLAHOMA CITY,OK,73132,1935-06-07,N080291233,NANCY JOAN COUCH,6612 RANDI RD,OKLAHOMA CITY,OK,73132,1935-06-07


### Add Similarity Scoring

In [84]:
!pip install rapidfuzz

In [85]:
from rapidfuzz import fuzz

def similarity(a, b):
    if pd.isna(a) or pd.isna(b):
        return 0
    return fuzz.WRatio(a, b)

block1_df["name_score"] = block1_df.apply(
    lambda r: similarity(r["kelmar_name"], r["sok_name"]),
    axis=1
)

block1_df["street_score"] = block1_df.apply(
    lambda r: similarity(r["kelmar_street"], r["sok_street"]),
    axis=1
)

block1_df["composite_score"] = (
    block1_df["name_score"] * 0.6 +
    block1_df["street_score"] * 0.4
)

block1_df.sort_values("composite_score", ascending=False).head(10)

,OwnerID,PropertyID,kelmar_name,kelmar_street,kelmar_city,kelmar_state,kelmar_zip,BirthDT,DLN,sok_name,sok_street,sok_city,sok_state,sok_zip,Date_of_Birth,name_score,street_score,composite_score
2,13439549,13099416,ALLEN H MOORE,2205 WINDSONG,EDMOND,OK,73034,1971-12-17,V081149038,ALLEN H MOORE,2205 WINDSONG,EDMOND,OK,73034,1971-12-17,100.0,100.0,100.0
10,12950318,12628565,CHRISTINE HENSHAW,10612 S MADISON ST,JENKS,OK,74037,1953-12-22,P080329904,CHRISTINE MARY HENSHAW,10612 S MADISON ST,JENKS,OK,74037,1953-12-22,95.0,100.0,97.0
16,12950494,12628734,ROBERT MORRISON,3709 S 65TH W AVE,TULSA,OK,74107,1921-05-31,F000041894,ROBERT A MORRISON,3709 S 65TH W AVE,TULSA,OK,74107,1921-05-31,95.0,100.0,97.0
4,12897104,12577135,NANCY COUCH,6612 RANDI RD,OKLAHOMA CITY,OK,73132,1935-06-07,N080291233,NANCY JOAN COUCH,6612 RANDI RD,OKLAHOMA CITY,OK,73132,1935-06-07,95.0,100.0,97.0
6,12950375,12628557,DIANNA TURNER,400482 W 2500 RD,BARTLESVILLE,OK,74006,1956-10-12,M082305662,DIANNA MAE TURNER,400482 W 2500 RD,BARTLESVILLE,OK,74006,1956-10-12,95.0,100.0,97.0
8,13595299,13246432,MEGAN HOLLOWAY,7217 E MEMPHIS ST,BROKEN ARROW,OK,74014,1983-07-24,K349716771,MEGAN DEANN HOLLOWAY,7217 E MEMPHIS ST,BROKEN ARROW,OK,74014,1983-07-24,95.0,100.0,97.0
9,12950506,12628644,ANDREW PRICE,312 E 151ST ST S,GLENPOOL,OK,74033,1978-02-14,F081208316,ANDREW NEAL PRICE,312 E 151ST ST S,GLENPOOL,OK,74033,1978-02-14,95.0,100.0,97.0
18,12950311,12628578,JOHN FULTZ,70200 S 340 AVE,WAGONER,OK,74467,1967-05-25,K081966024,JOHN RAY FULTZ,70200 S 340 AVE,WAGONER,OK,74467,1967-05-25,95.0,100.0,97.0
12,12950357,12628512,CARRIE PITTS,6502 S MEADOWOOD DR,OWASSO,OK,74055,1969-11-11,M701621668,CARRIE ANN PITTS,6502 S MEADOWOOD DR,OWASSO,OK,74055,1969-11-11,95.0,100.0,97.0
13,12950391,12628602,JOHNNIE BYROM,23991 W 102ND ST S,SAPULPA,OK,74066,1952-07-03,K082090009,JOHNNIE VANCE BYROM,23991 W 102ND ST S,SAPULPA,OK,74066,1952-07-03,95.0,100.0,97.0


In [86]:
block1_df.sort_values("composite_score", ascending=False).tail(10)

,OwnerID,PropertyID,kelmar_name,kelmar_street,kelmar_city,kelmar_state,kelmar_zip,BirthDT,DLN,sok_name,sok_street,sok_city,sok_state,sok_zip,Date_of_Birth,name_score,street_score,composite_score
5,12896165,12582627,TERRY PAYNE,626 STONERIDGE DR,MOORE,OK,73160,1963-07-26,N082578316,TERRY JANE PAYNE,626 STONERIDGE,MOORE,OK,73160,1963-07-26,95.000000,95.000000,95.000000
7,12950466,12628726,BARBARA HELMER,636 SE ELMHURST AVE,BARTLESVILLE,OK,74006,1932-01-01,B100026297,BARBARA JEAN HELMER,636 SE ELMHURST,BARTLESVILLE,OK,74006,1932-01-01,95.000000,95.000000,95.000000
3,13440139,13100006,ROBERT N ALLEN,2101 N FONSHILL AVE,OKLAHOMA CITY,OK,73111,1946-01-22,447449471,ROBERT NELSON ALLEN,2101 N FONSHILL AVE,OKLA CITY,OK,73111,1946-01-22,87.692308,100.000000,92.615385
11,12950335,12628538,STEPHANIE MEKUSKY,8748 N MINGO RD APT 315,OWASSO,OK,74055,1961-03-27,K083238643,STEPHANIE ANNE MEKUSKY,8748 N MINGO RD APT 808,OWASSO,OK,74055,1961-03-27,95.000000,86.956522,91.782609
14,12950451,12628698,ROBERT FORD,1209 E 5TH ST,STILLWATER,OK,74074,1975-12-12,P082635444,ROBERT W FORD,1209 E 5TH AVE,STILLWATER,OK,74074,1975-12-12,95.000000,82.608696,90.043478
17,12950462,12628654,ROBERT HARING,1620 S DETROIT,TULSA,OK,74120,1932-11-13,P081361922,ROBERT WESTING HARING,1620 S DETROIT AVE,TULSA,OK,74120,1932-11-13,85.500000,95.000000,89.300000
19,12896242,12583698,SANDY J SANFORD,2020 WOODLAND DR,ADA,OK,74820,1943-10-09,V001545330,CASONDRA JEAN SANFORD,2020 WOODLAND DR,ADA,OK,74820,1943-10-09,72.222222,100.000000,83.333333
15,12950343,12628576,LULA MUNION,3141 SO LEWIS AVE,TULSA,OK,74105,1934-10-27,T080996495,FOTINI MUNION,3141 S LEWIS AVE,TULSA,OK,74105,1934-10-27,67.058824,96.969697,79.023173
1,12771950,12466480,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,507 W MAIN ST,CALUMET,OK,73014,1960-05-07,89.473684,18.181818,60.956938
0,12771949,12466462,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,507 W MAIN ST,CALUMET,OK,73014,1960-05-07,89.473684,18.181818,60.956938


In [87]:
block1_df["composite_score"].describe()

count     20.000000
mean      90.550593
std       11.349428
min       60.956938
25%       89.857609
50%       96.000000
75%       97.000000
max      100.000000
Name: composite_score, dtype: float64

In [89]:
import numpy as np

block1_df["bucket"] = np.where(
    block1_df["composite_score"] >= 92, "AUTO_APPROVE",
    np.where(block1_df["composite_score"] >= 85, "REVIEW", "REJECT")
)

block1_df["bucket"].value_counts()

bucket
AUTO_APPROVE    13
REJECT           4
REVIEW           3
Name: count, dtype: int64

In [90]:
from datetime import datetime

block1_df["block_name"] = "BLOCK1_ZIP_LAST_DOB"
block1_df["match_stage"] = "FUZZY"
block1_df["scoring_version"] = "v1"
block1_df["created_at"] = datetime.utcnow()

block1_df.head()

,OwnerID,PropertyID,kelmar_name,kelmar_street,kelmar_city,kelmar_state,kelmar_zip,BirthDT,DLN,sok_name,...,sok_zip,Date_of_Birth,name_score,street_score,composite_score,bucket,block_name,match_stage,scoring_version,created_at
0,12771949,12466462,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,...,73014,1960-05-07,89.473684,18.181818,60.956938,REJECT,BLOCK1_ZIP_LAST_DOB,FUZZY,v1,2026-03-03 16:20:59.301545
1,12771950,12466480,DEBORAH R WOODMAN,PO BOX 33,CALUMET,OK,73014,1960-05-07,D391345720,DEBORAH RENNE WOODMAN,...,73014,1960-05-07,89.473684,18.181818,60.956938,REJECT,BLOCK1_ZIP_LAST_DOB,FUZZY,v1,2026-03-03 16:20:59.301545
2,13439549,13099416,ALLEN H MOORE,2205 WINDSONG,EDMOND,OK,73034,1971-12-17,V081149038,ALLEN H MOORE,...,73034,1971-12-17,100.000000,100.000000,100.000000,AUTO_APPROVE,BLOCK1_ZIP_LAST_DOB,FUZZY,v1,2026-03-03 16:20:59.301545
3,13440139,13100006,ROBERT N ALLEN,2101 N FONSHILL AVE,OKLAHOMA CITY,OK,73111,1946-01-22,447449471,ROBERT NELSON ALLEN,...,73111,1946-01-22,87.692308,100.000000,92.615385,AUTO_APPROVE,BLOCK1_ZIP_LAST_DOB,FUZZY,v1,2026-03-03 16:20:59.301545
4,12897104,12577135,NANCY COUCH,6612 RANDI RD,OKLAHOMA CITY,OK,73132,1935-06-07,N080291233,NANCY JOAN COUCH,...,73132,1935-06-07,95.000000,100.000000,97.000000,AUTO_APPROVE,BLOCK1_ZIP_LAST_DOB,FUZZY,v1,2026-03-03 16:20:59.301545


In [91]:
BLOCK1_RESULTS = f"{PROJECT_ID}.{DATASET}.fuzzy_block1_classified_v1"

block1_df.to_gbq(
    BLOCK1_RESULTS,
    project_id=PROJECT_ID,
    if_exists="replace"
)

print("✅ Block 1 classified results saved:", len(block1_df))

/var/tmp/ipykernel_8090/1882096400.py:3: FutureWarning: to_gbq is deprecated and will be removed in a future version. Please use pandas_gbq.to_gbq instead: https://pandas-gbq.readthedocs.io/en/latest/api.html#pandas_gbq.to_gbq
  block1_df.to_gbq(
100%|██████████| 1/1 [00:00<00:00, 8886.24it/s]

✅ Block 1 classified results saved: 20


### Block 2: ZIP + Last Name
(no DOB requirement)

In [92]:
BLOCK2 = f"{PROJECT_ID}.{DATASET}.fuzzy_block2_candidates_v1"

query = f"""
CREATE OR REPLACE TABLE `{BLOCK2}` AS
SELECT
    k.OwnerID,
    k.PropertyID,
    k.full_name_clean AS kelmar_name,
    k.street_clean AS kelmar_street,
    k.city_clean AS kelmar_city,
    k.state_clean AS kelmar_state,
    k.zip_clean AS kelmar_zip,
    k.BirthDT,

    s.DLN,
    s.full_name_clean AS sok_name,
    s.street_clean AS sok_street,
    s.city_clean AS sok_city,
    s.state_clean AS sok_state,
    s.zip_clean AS sok_zip,
    s.Date_of_Birth

FROM `{PROJECT_ID}.{DATASET}.kelmar_fuzzy_v1` k
JOIN `{PROJECT_ID}.{DATASET}.sok_fuzzy_pool_v1` s
ON k.zip_clean = s.zip_clean
AND k.last_name_clean = s.last_name_clean
"""

bq.query(query).result()

print("✅ Block 2 candidate table created:", BLOCK2)

✅ Block 2 candidate table created: omes-solacc-citizenmatch-01-d.citizen_match.fuzzy_block2_candidates_v1


In [93]:
print(
    bq.query(f"SELECT COUNT(*) AS candidate_count FROM `{BLOCK2}`")
    .to_dataframe()
)

   candidate_count
0            27455


In [94]:
BLOCK2 = f"{PROJECT_ID}.{DATASET}.fuzzy_block2_candidates_v1"

block2_df = bq.query(f"SELECT * FROM `{BLOCK2}`").to_dataframe()
print("Loaded Block 2 candidates:", block2_df.shape)

Loaded Block 2 candidates: (27455, 15)


In [95]:
from rapidfuzz import fuzz

def similarity(a, b):
    if pd.isna(a) or pd.isna(b):
        return 0
    return fuzz.WRatio(a, b)

block2_df["name_score"] = block2_df.apply(
    lambda r: similarity(r["kelmar_name"], r["sok_name"]),
    axis=1
)

block2_df["street_score"] = block2_df.apply(
    lambda r: similarity(r["kelmar_street"], r["sok_street"]),
    axis=1
)

block2_df["composite_score"] = (
    block2_df["name_score"] * 0.6 +
    block2_df["street_score"] * 0.4
)

In [104]:
import numpy as np
from datetime import datetime

block2_df["bucket"] = np.where(
    block2_df["composite_score"] >= 94, "AUTO_APPROVE",
    np.where(block2_df["composite_score"] >= 88, "REVIEW", "REJECT")
)

block2_df["block_name"] = "BLOCK2_ZIP_LAST"
block2_df["match_stage"] = "FUZZY"
block2_df["scoring_version"] = "v1"
block2_df["created_at"] = datetime.utcnow()

block2_df["bucket"].value_counts()

bucket
REJECT          27006
REVIEW            284
AUTO_APPROVE      165
Name: count, dtype: int64

In [105]:
BLOCK2_RESULTS = f"{PROJECT_ID}.{DATASET}.fuzzy_block2_classified_v1"

block2_df.to_gbq(
    BLOCK2_RESULTS,
    project_id=PROJECT_ID,
    if_exists="replace"
)

print("✅ Block 2 classified results saved:", len(block2_df))

/var/tmp/ipykernel_8090/842846427.py:3: FutureWarning: to_gbq is deprecated and will be removed in a future version. Please use pandas_gbq.to_gbq instead: https://pandas-gbq.readthedocs.io/en/latest/api.html#pandas_gbq.to_gbq
  block2_df.to_gbq(
100%|██████████| 1/1 [00:00<00:00, 14266.34it/s]

✅ Block 2 classified results saved: 27455


In [106]:
auto_block1 = block1_df[block1_df["bucket"] == "AUTO_APPROVE"][["OwnerID","DLN"]]
auto_block2 = block2_df[block2_df["bucket"] == "AUTO_APPROVE"][["OwnerID","DLN"]]

combined_auto = pd.concat([auto_block1, auto_block2]).drop_duplicates()

print("Unique fuzzy auto matches:", combined_auto.shape[0])

Unique fuzzy auto matches: 164


### Final Table

In [120]:
import pandas as pd
from google.cloud import bigquery

# -------------------
# Config
# -------------------
PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

FINAL_TABLE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"

SOK_TOKEN = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset_tokenized_v2"
SOK_CLEAN = f"{PROJECT_ID}.{DATASET}.sok_clean_v1"
KELMAR_CLEAN = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"

BLOCK1_TBL = f"{PROJECT_ID}.{DATASET}.fuzzy_block1_classified_v1"
BLOCK2_TBL = f"{PROJECT_ID}.{DATASET}.fuzzy_block2_classified_v1"

DET_TBL = f"{PROJECT_ID}.{DATASET}.ssn_deterministic_matches_v1"

bq = bigquery.Client(project=PROJECT_ID)

# =========================================================
# 0) Build DLN -> Transaction mapping (latest per DLN)
# =========================================================
# We select ONE "canonical" transaction row per DLN for lineage.
# Latest _data_file_date_ wins; then latest Transaction_Date; then non-empty Transaction_ID preferred.
dln_map = bq.query(f"""
SELECT
  CAST(DLN AS STRING) AS DLN,
  NULLIF(CAST(Transaction_ID AS STRING), '') AS Transaction_ID,
  _data_file_date_
FROM `{SOK_TOKEN}`
WHERE DLN IS NOT NULL
QUALIFY ROW_NUMBER() OVER (
  PARTITION BY CAST(DLN AS STRING)
  ORDER BY
    _data_file_date_ DESC,
    SAFE_CAST(Transaction_Date AS TIMESTAMP) DESC,
    IF(NULLIF(CAST(Transaction_ID AS STRING), '') IS NULL, 1, 0) ASC
) = 1
""").to_dataframe()

print("DLN map rows:", dln_map.shape)

# =========================================================
# 1) Deterministic matches (join back for full fields + lineage)
# =========================================================
det_full = bq.query(f"""
SELECT
    d.OwnerID,
    d.PropertyID,
    CAST(d.DLN AS STRING) AS DLN,

    k.full_name_clean AS kelmar_name,
    k.street_clean AS kelmar_street,
    k.city_clean AS kelmar_city,
    k.state_clean AS kelmar_state,
    k.zip_clean AS kelmar_zip,
    k.BirthDT,

    sc.full_name_clean AS sok_name,
    sc.street_clean AS sok_street,
    sc.city_clean AS sok_city,
    sc.state_clean AS sok_state,
    sc.zip_clean AS sok_zip,
    sc.Date_of_Birth,

    'DETERMINISTIC_SSN' AS technique,
    100.0 AS name_score,
    100.0 AS street_score,
    100.0 AS composite_score,
    100.0 AS confidence_score,
    'AUTO_APPROVE' AS bucket

FROM `{DET_TBL}` d
JOIN `{KELMAR_CLEAN}` k
  ON d.OwnerID = k.OwnerID AND d.PropertyID = k.PropertyID
JOIN `{SOK_CLEAN}` sc
  ON CAST(d.DLN AS STRING) = CAST(sc.DLN AS STRING)
""").to_dataframe()

# Attach transaction lineage via DLN map
det_full = det_full.merge(dln_map, on="DLN", how="left")

print("Deterministic rows:", det_full.shape[0])

# =========================================================
# 2) Load fuzzy block tables (they already have DLN + scores + buckets)
# =========================================================
block1 = bq.query(f"SELECT * FROM `{BLOCK1_TBL}`").to_dataframe()
block2 = bq.query(f"SELECT * FROM `{BLOCK2_TBL}`").to_dataframe()

# Ensure DLN is string
block1["DLN"] = block1["DLN"].astype("string")
block2["DLN"] = block2["DLN"].astype("string")

# Technique + confidence
block1["technique"] = "ZIP+LAST+DOB_STRONG_BLOCK"
block1["confidence_score"] = 95 * (block1["composite_score"] / 100)

block2["technique"] = "ZIP+LAST_FUZZY_BLOCK"
block2["confidence_score"] = 85 * (block2["composite_score"] / 100)

# Attach transaction lineage via DLN map
block1 = block1.merge(dln_map, on="DLN", how="left")
block2 = block2.merge(dln_map, on="DLN", how="left")

# =========================================================
# 3) Standardize column set (including transaction lineage)
# =========================================================
base_cols = [
    "OwnerID","PropertyID","DLN",
    "Transaction_ID","_data_file_date_",
    "kelmar_name","kelmar_street","kelmar_city","kelmar_state","kelmar_zip","BirthDT",
    "sok_name","sok_street","sok_city","sok_state","sok_zip","Date_of_Birth",
    "technique","name_score","street_score","confidence_score","composite_score","bucket"
]

# Some tables might not have exact dtypes; enforce presence
for df_name, df in [("det_full", det_full), ("block1", block1), ("block2", block2)]:
    missing = [c for c in base_cols if c not in df.columns]
    if missing:
        raise ValueError(f"{df_name} missing columns: {missing}")

det_full = det_full[base_cols]
block1 = block1[base_cols]
block2 = block2[base_cols]

# =========================================================
# 4) Combine + dedupe (keep highest confidence per OwnerID+PropertyID+DLN)
# =========================================================
combined = pd.concat([det_full, block1, block2], ignore_index=True)
print("Combined before dedupe:", combined.shape[0])

combined = (
    combined
    .sort_values("confidence_score", ascending=False)
    .drop_duplicates(subset=["OwnerID","PropertyID","DLN"])
)

print("Final unique proposals:", combined.shape[0])

# =========================================================
# 5) Persist to BigQuery
# =========================================================
combined.to_gbq(
    FINAL_TABLE,
    project_id=PROJECT_ID,
    if_exists="replace"
)

print("✅ FINAL Treasury table created:", FINAL_TABLE)
print("Total rows delivered:", combined.shape[0])

DLN map rows: (4030940, 3)
Deterministic rows: 3562
Combined before dedupe: 31037
Final unique proposals: 30617


/var/tmp/ipykernel_8090/2336876738.py:147: FutureWarning: to_gbq is deprecated and will be removed in a future version. Please use pandas_gbq.to_gbq instead: https://pandas-gbq.readthedocs.io/en/latest/api.html#pandas_gbq.to_gbq
  combined.to_gbq(
100%|██████████| 1/1 [00:00<00:00, 14873.42it/s]

✅ FINAL Treasury table created: omes-solacc-citizenmatch-01-d.citizen_match.treasury_match_review_v2
Total rows delivered: 30617


### Deterministic always included (high confidence)

Fuzzy limited to top 3

No candidate explosion

Still transparent

Treasury can sort by confidence

Note: 4,472 properties with candidates

528 properties with zero candidates

In [129]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

SOURCE_TABLE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"
FINAL_CAPPED = f"{PROJECT_ID}.{DATASET}.treasury_match_review_capped_v1"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
CREATE OR REPLACE TABLE `{FINAL_CAPPED}` AS

WITH ranked AS (
  SELECT
    *,
    ROW_NUMBER() OVER (
      PARTITION BY OwnerID
      ORDER BY confidence_score DESC
    ) AS rank_within_owner
  FROM `{SOURCE_TABLE}`
)

SELECT *
FROM ranked
WHERE
  technique = 'DETERMINISTIC_SSN'
  OR rank_within_owner <= 3
"""

bq.query(query).result()

print("✅ Final capped Treasury table created:", FINAL_CAPPED)

✅ Final capped Treasury table created: omes-solacc-citizenmatch-01-d.citizen_match.treasury_match_review_capped_v1


In [130]:
bq.query(f"""
SELECT COUNT(*) AS total_rows
FROM `{FINAL_CAPPED}`
""").to_dataframe()

,total_rows
0,6005


In [131]:
bq.query(f"""
SELECT COUNT(DISTINCT OwnerID) AS owner_count
FROM `{FINAL_CAPPED}`
""").to_dataframe()

,owner_count
0,4472


### Build Unmatched Table

In [132]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

TREASURY_REVIEW = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"
KELMAR_CLEAN = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"

TREASURY_UNMATCHED = f"{PROJECT_ID}.{DATASET}.treasury_unmatched_v1"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
CREATE OR REPLACE TABLE `{TREASURY_UNMATCHED}` AS
SELECT
    k.OwnerID,
    k.PropertyID,
    k.full_name_clean AS kelmar_name,
    k.street_clean AS kelmar_street,
    k.city_clean AS kelmar_city,
    k.state_clean AS kelmar_state,
    k.zip_clean AS kelmar_zip,
    k.BirthDT
FROM `{KELMAR_CLEAN}` k
LEFT JOIN `{TREASURY_REVIEW}` r
    ON k.OwnerID = r.OwnerID
WHERE r.OwnerID IS NULL
"""

bq.query(query).result()

print("✅ Treasury unmatched table created:", TREASURY_UNMATCHED)

✅ Treasury unmatched table created: omes-solacc-citizenmatch-01-d.citizen_match.treasury_unmatched_v1


In [133]:
bq.query(f"""
SELECT COUNT(*) AS unmatched_count
FROM `{TREASURY_UNMATCHED}`
""").to_dataframe()

,unmatched_count
0,528


### Validate Kelmar values are unchanged
OwnerID exists in original Kelmar table

PropertyID exists

BirthDT matches

Name fields match original cleaned version

In [134]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

TREASURY_TABLE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"
KELMAR_CLEAN = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
SELECT
    COUNT(*) AS total_rows_checked,

    COUNTIF(k.OwnerID IS NULL) AS missing_ownerid,
    COUNTIF(k.PropertyID IS NULL) AS missing_propertyid,

    COUNTIF(t.BirthDT != k.BirthDT) AS birthdt_mismatch,

    COUNTIF(t.kelmar_name != k.full_name_clean) AS name_mismatch,
    COUNTIF(t.kelmar_street != k.street_clean) AS street_mismatch,
    COUNTIF(t.kelmar_city != k.city_clean) AS city_mismatch,
    COUNTIF(t.kelmar_state != k.state_clean) AS state_mismatch,
    COUNTIF(t.kelmar_zip != k.zip_clean) AS zip_mismatch

FROM `{TREASURY_TABLE}` t
LEFT JOIN `{KELMAR_CLEAN}` k
  ON t.OwnerID = k.OwnerID
 AND t.PropertyID = k.PropertyID
"""

result = bq.query(query).to_dataframe()
result

,total_rows_checked,missing_ownerid,missing_propertyid,birthdt_mismatch,name_mismatch,street_mismatch,city_mismatch,state_mismatch,zip_mismatch
0,30617,0,0,0,0,0,0,0,0


### Validate SOK values are traceable

In [135]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

TREASURY_TABLE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"
SOK_CLEAN = f"{PROJECT_ID}.{DATASET}.sok_clean_v1"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
SELECT
    COUNT(*) AS total_rows_checked,

    COUNTIF(s.DLN IS NULL) AS missing_dln,

    COUNTIF(t.sok_name != s.full_name_clean) AS name_mismatch,
    COUNTIF(t.sok_street != s.street_clean) AS street_mismatch,
    COUNTIF(t.sok_city != s.city_clean) AS city_mismatch,
    COUNTIF(t.sok_state != s.state_clean) AS state_mismatch,
    COUNTIF(t.sok_zip != s.zip_clean) AS zip_mismatch,

    COUNTIF(t.Date_of_Birth != s.Date_of_Birth) AS dob_mismatch

FROM `{TREASURY_TABLE}` t
LEFT JOIN `{SOK_CLEAN}` s
  ON t.DLN = s.DLN
"""

result = bq.query(query).to_dataframe()
result

,total_rows_checked,missing_dln,name_mismatch,street_mismatch,city_mismatch,state_mismatch,zip_mismatch,dob_mismatch
0,31039,1477,4,122,46,0,76,0


### Expanding table columns
CashValue → from kelmar_clean_v1

Deceased → from SOK source tables

In [144]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

SOURCE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v2"
FINAL  = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v3"

KELMAR_CLEAN = f"{PROJECT_ID}.{DATASET}.kelmar_clean_v1"
SOK_STAGING  = f"{PROJECT_ID}.{DATASET}.sok_staging_dataset"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
CREATE OR REPLACE TABLE `{FINAL}` AS

WITH deceased_person AS (
  SELECT
    CAST(DLN AS STRING) AS DLN,
    ANY_VALUE(Deceased) AS Deceased
  FROM `{SOK_STAGING}`
  WHERE DLN IS NOT NULL
  GROUP BY DLN
)

SELECT
    t.*,
    k.CashValue,
    d.Deceased

FROM `{SOURCE}` t

LEFT JOIN `{KELMAR_CLEAN}` k
  ON t.OwnerID = k.OwnerID
 AND t.PropertyID = k.PropertyID

LEFT JOIN deceased_person d
  ON t.DLN = d.DLN
"""

bq.query(query).result()

print("✅ treasury_match_review_v3")

✅ treasury_match_review_v3


In [145]:
bq.query(f"""
SELECT COUNT(*) AS total_rows
FROM `{FINAL}`
""").to_dataframe()

,total_rows
0,30617


In [146]:
from google.cloud import bigquery

PROJECT_ID = "omes-solacc-citizenmatch-01-d"
DATASET = "citizen_match"

SOURCE = f"{PROJECT_ID}.{DATASET}.treasury_match_review_v3"
FINAL  = f"{PROJECT_ID}.{DATASET}.treasury_match_review_capped_v2"

bq = bigquery.Client(project=PROJECT_ID)

query = f"""
CREATE OR REPLACE TABLE `{FINAL}` AS

WITH enriched AS (
  SELECT
    *,
    
    -- Eligibility flag (person-level)
    CASE 
      WHEN Deceased = TRUE THEN 'INELIGIBLE_DECEASED'
      ELSE 'ELIGIBLE'
    END AS Eligibility_Flag,

    -- Rank within OwnerID by confidence
    ROW_NUMBER() OVER (
      PARTITION BY OwnerID
      ORDER BY confidence_score DESC
    ) AS owner_rank

  FROM `{SOURCE}`
)

SELECT *
FROM enriched
WHERE
  -- Keep ALL deterministic matches
  technique = 'DETERMINISTIC_SSN'
  
  -- Keep top 3 per OwnerID for fuzzy
  OR owner_rank <= 3
"""

bq.query(query).result()

print("✅ treasury_match_review_capped_v2 created successfully")

✅ treasury_match_review_capped_v2 created successfully
